# Descriptive Statistics with Pandas

### Objectives
* Read data from .csv files into `pandas` Dataframes  
* Slice dataframes for basic queries  
* Find Measures of Center and demonstrate understanding of specific use cases for each
* Find Measures of Dispersion and demonstrate understanding of specific use cases for each
* Manipulate data to find answers to questions using Python and `pandas`



## The dataset

For this lab, we'll be exploring the [Titanic Dataset](https://www.kaggle.com/c/titanic).  This dataset contains records from a subset of passengers that were aboard the Titanic when it sank. (Spoiler Alert, if you haven't seen the movie yet). 

In this notebook, we're going to explore the dataset and see if we can find any trends in the data that suggest if there were certain traits that made some passengers more (or less) likely to survive than others.  
<br>  
<center><img src='titanic-door.jpeg'>
_Data shows that having a partner that is a door hog negatively affects a passenger's chances of survival._ 
</center>

## Working with Data in Pandas
<center>
    <br>
In order to make working with the data easy, we're going to use every Data Scientist's favorite tool, **_[Pandas](https://pandas.pydata.org/pandas-docs/stable/)_**!


<img src='pandas-dj.gif'>
</center>

Pandas is perhaps the most widely used python library in all of Data Science.  This powerful library makes it easy for us to read in, view, and manipulate datasets by using **_Dataframes_**.  

We'll start by importing the library, as well as reading in the dataset and storing it as a Dataframe.  

Since we'll be using the `pandas` library a lot, it's common to see people alias it as `pd`, to save some keystrokes every time we need to access it. 

In the cell below, complete the following steps:

1.  import pandas and set an alias by typing `import pandas as pd`.   
<br>  
1. Use the pandas function `read_csv()` to read in the dataset stored in `titanic.csv`.  To do this, pass the name of the file in to `pd.read_csv()` as a string.  Store the dataframe that this function returns in the `df` variable.  
<br>  
1. Display the dataframe by typing `df` at the bottom of the cell before running it.  This will cause the Jupyter Notebook to display the contents of the df variable, which is our dataframe.  

In [1]:
import pandas as pd
df = pd.read_csv('titanic.csv')
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
5              6         0       3   
6              7         0       1   
7              8         0       3   
8              9         1       3   
9             10         1       2   
10            11         1       3   
11            12         1       1   
12            13         0       3   
13            14         0       3   
14            15         0       3   
15            16         1       2   
16            17         0       3   
17            18         1       2   
18            19         0       3   
19            20         1       3   
20            21         0       2   
21            22         1       2   
22            23         1       3   
23            24         1       1   
24            25         0       3   
25            26         1       3   
26            27         0       3   
27            28         0       1   
28            29         1       3   
29            30         0       3   
..           ...       ...     ...   
861          862         0       2   
862          863         1       1   
863          864         0       3   
864          865         0       2   
865          866         1       2   
866          867         1       2   
867          868         0       1   
868          869         0       3   
869          870         1       3   
870          871         0       3   
871          872         1       1   
872          873         0       1   
873          874         0       3   
874          875         1       2   
875          876         1       3   
876          877         0       3   
877          878         0       3   
878          879         0       3   
879          880         1       1   
880          881         1       2   
881          882         0       3   
882          883         0       3   
883          884         0       2   
884          885         0       3   
885          886         0       3   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
5                                     Moran, Mr. James    male   NaN      0   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
7                       Palsson, Master. Gosta Leonard    male   2.0      3   
8    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                  Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
10                     Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
12                      Saundercock, Mr. William Henry    male  20.0      0   
13                         Andersson, Mr. Anders Johan    male  39.0      1   
14                Vestrom, Miss. Hulda Amanda Adolfina  female  14.0      0   
15                    Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
16                                Rice, Master. Eugene    male   2.0      4   
17                        Williams, Mr. Charles Eugene    male   NaN      0   
18   Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0      1   
19                             Masselmani, Mrs. Fatima  female   NaN

Great! That was easy.  If we scroll to the bottom of the dataframe displayed above, we can see that there are records for 890 passengers.  

For convenience's sake, we've included the **_Data Dictionary_** from Kaggle for this dataset.  Data dictionaries are helpful tools that allow us to make sense of the data we're looking at. <table>
<tbody>
<tr><th><b>Variable</b></th><th><b>Definition</b></th><th><b>Key</b></th></tr>
<tr>
<td>survival</td>
<td>Survival</td>
<td>0 = No, 1 = Yes</td>
</tr>
<tr>
<td>pclass</td>
<td>Ticket class</td>
<td>1 = 1st, 2 = 2nd, 3 = 3rd</td>
</tr>
<tr>
<td>sex</td>
<td>Sex</td>
<td></td>
</tr>
<tr>
<td>Age</td>
<td>Age in years</td>
<td></td>
</tr>
<tr>
<td>sibsp</td>
<td># of siblings / spouses aboard the Titanic</td>
<td></td>
</tr>
<tr>
<td>parch</td>
<td># of parents / children aboard the Titanic</td>
<td></td>
</tr>
<tr>
<td>ticket</td>
<td>Ticket number</td>
<td></td>
</tr>
<tr>
<td>fare</td>
<td>Passenger fare</td>
<td></td>
</tr>
<tr>
<td>cabin</td>
<td>Cabin number</td>
<td></td>
</tr>
<tr>
<td>embarked</td>
<td>Port of Embarkation</td>
<td>C = Cherbourg, Q = Queenstown, S = Southampton</td>
</tr>
</tbody>
</table>

See if you can use the Data Dictionary to answer the following questions:

* What does the `PClass` column tell us? 
* In the `Embarked` column, what does a value of `S` mean?
* In the `Survived` column, does a `1` mean the passenger died, or they survived?

With a data dictionary, answering these questions is easy.  Without one, however, it would be pretty hard to figure this out, and would probably require more time Googling then we would like to spend.  


Typically, when loading in a new data set, it's a good idea to look at a selection of the data, and consult the Data Dictionary to confirm that you understand what you're looking at.  

If you want to look at the first or last few rows in a dataframe without printing out the rest of the data, `pandas` makes that easy, too.  

In the cell below: 

1.  Call the `df` object's `.head()` method to print out the first 5 rows of the dataframe.  

In [2]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

Great! Now that we know how to get data into a dataframe and display it, we're ready to start digging into the statistics of it all.  

### Summary Statistics in Pandas

In this section, we'll explore the dataframe as a whole, and then start looking for interesting subgroups to check out.  During this exploration, we'll rely heavily on **_Descriptive Statistics_**.  Let's review some of the more important concepts to make sure we're up to speed on them.  They can be divided into two subgroups: **_Measures of Center_** and **_Measures of Spread_**.

#### Measures of Center

**_Mean_**:  Also called the average.  The arithmetic mean of a set--the sum of the data points in a set, divided by the number of data points.  The mean is heavily affected by outliers (extremely large or small values that are far away from the majority of points in the dataset)

**_Median_**: The middle value in a dataset.  If the values were sorted and broken into two sets of equal length,  the median value would fall directly in the middle.  The median is less affected by outliers, since they can only shift the midpoint of the sorted dataset by 1 value.  

**_Mode_**: The number (or numbers) that appear the most in a dataset.  The mode is not affected by outliers at all.

#### Measures of Spread

**_Standard Deviation_**: How far the typical value tends to be from the mean. Standard Deviation is the square root of Variance. 

**_Variance_**: How far a set of of numbers are spread out from the mean.  Variance is the Square of Standard Deviation.  

#### Others Worth Knowing

**_Minimum/Maximum_**: The smallest/largest values in a set.  These two are pretty self-explanatory. 

**_Quartiles/Quantiles_**:  Quartiles are a special case of quantiles.  Much like the median, quartiles are when the data is sorted and divided into equal portions. In the case of quartiles, these would be 4 equal portions, although the term quantiles can be used to describe any arbitrary number of splits.  The quartile values would be the values that the dataset splits on--the value at 25%, 50%, and 75%.

Let's put our dataframe to use and take a look at the descriptive statistics for this dataset.

In the cell below:

1. Call the `df` object's `.describe()` method.

In [3]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

Take a look at the table above.  See if you can answer the following questions:

1.  What is the average age of a passenger on the titanic?  What is the Standard Deviation for this value, and what does that mean?

1. What is the most expensive price paid for a ticket on the titanic? What is the least expensive ticket price?

1.  What is the median ticket price?


We can also use the `.describe()` function on a single column.  For instance, if we wanted to get the descriptive statistics on the `Fare` column, we could type:
```python
df['Fare'].describe()
```

In the cell below, call the `.describe()` method on the `Age` column. 

In [4]:
df['Age'].describe()

count    714.000000
mean      29.699118
std       14.526497
min        0.420000
25%       20.125000
50%       28.000000
75%       38.000000
max       80.000000
Name: Age, dtype: float64

If we want to view more than 1 column at the same time, we can pass a list containing the column names into in the square brackets as the index.

For instance:

```python
df[['Pclass', 'Age']].describe()
```

prints out the descriptive statistics for both columns.  

In the cell below, display the descriptive statistics for only the `Age` and `Fare` columns.  

In [5]:
df[['Age', 'Fare']].describe()

Age        Fare
count  714.000000  891.000000
mean    29.699118   32.204208
std     14.526497   49.693429
min      0.420000    0.000000
25%     20.125000    7.910400
50%     28.000000   14.454200
75%     38.000000   31.000000
max     80.000000  512.329200

### Dataframe Slicing in Pandas

Descriptive statistics on the entire dataframe can only tell us so much. For our next step, we'll **_slice_** our dataframe into interesting subgroups, and see if we can find anything interesting.  Let's start by slicing the most obvious group in our dataset--the survivors. The syntax here is a bit weird, so we'll do this in stages.  

Run the cell below to see what it returns:

In [6]:
df['Survived'] == 1

0      False
1       True
2       True
3       True
4      False
5      False
6      False
7      False
8       True
9       True
10      True
11      True
12     False
13     False
14     False
15      True
16     False
17      True
18     False
19      True
20     False
21      True
22      True
23      True
24     False
25      True
26     False
27     False
28      True
29     False
       ...  
861    False
862     True
863    False
864    False
865     True
866     True
867    False
868    False
869     True
870    False
871     True
872    False
873    False
874     True
875     True
876    False
877    False
878    False
879     True
880     True
881    False
882    False
883    False
884    False
885    False
886    False
887     True
888    False
889     True
890    False
Name: Survived, Length: 891, dtype: bool

As we can see from the output above, using the `==` operator just returns a list of booleans for that column, which specifies which values meet that criteria and which don't.  If we want a dataframe that contains the entire row data for every row where `Survived` equals 1, we need to wrap the statement above in another slice operation.  

For instance, if we wanted to slice the dataframe and get only the rows where `Pclass` == 1, we would type:

```python
df[df["Pclass"] == 1]
```

In the cell below:

1. Slice the dataframe so that it contains only passengers who survived.  Store the dataframe that is returned in the `survivors_df` variable. 

2. Use `.head()` to display the first 5 rows from `survivors_df`

In [7]:
# Step 1: Slice a dataframe that only contains the Surviving passengers
survivors_df = df[df['Survived'] == 1]

# Step 2: Display the head of this new dataframe. 
survivors_df.head()

PassengerId  Survived  Pclass  \
1            2         1       1   
2            3         1       3   
3            4         1       1   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

We can also chain statements like the above together to slice on conditional statements using `and` and `or`.  

For instance, if we wanted a dataframe that only contained female passengers that didn't survive, we would type:

```python
df[(df["Sex"] == 'female') & (df['Survived'] == 0)]
```

Note that the syntax for this sort of slicing operation differs slightly from vanilla python.  Above, we had to use `&` to represent `and`.  When we want to use `|` for `or`.

in the cell below:

1. Slice a dataframe that contains only male passengers over the age of 18 with 1 or more children (refer to the data dictionary if you're unsure how this is represented). Store this dataframe in `dad_df`.  

1. Print out the head of `dad_df`

In [8]:
# Step 1: Slice a dataframe that contains only male passengers with 1 or more chidren.  
dad_df = df[(df["Sex"] == 'male') & (df['Parch'] > 0) & (df['Age'] >= 18)]

# Step 2: Print the head of dad_df
dad_df.head()

PassengerId  Survived  Pclass                             Name   Sex  \
13           14         0       3      Andersson, Mr. Anders Johan  male   
27           28         0       1   Fortune, Mr. Charles Alexander  male   
54           55         0       1   Ostby, Mr. Engelhart Cornelius  male   
93           94         0       3          Dean, Mr. Bertram Frank  male   
97           98         1       1  Greenfield, Mr. William Bertram  male   

     Age  SibSp  Parch     Ticket      Fare        Cabin Embarked  
13  39.0      1      5     347082   31.2750          NaN        S  
27  19.0      3      2      19950  263.0000  C23 C25 C27        S  
54  65.0      0      1     113509   61.9792          B30        C  
93  26.0      1      2  C.A. 2315   20.5750          NaN        S  
97  23.0      0      1   PC 17759   63.3583      D10 D12        C

## Putting it All Together

Now that we've explored the basic skills necessary to explore a dataset, let's put it all together and see if we can find some interesting trends in the data.  

In whatever way you see fit, find answers to the following questions:

1. What was is the median `Parch` value for men that survived the titanic disaster? What about the mean ticket price of men that survived?   
<br>    
1. Of all the women that did not survive, what percentage have a `Pclass` value of 2 or 3?  
<br>  
1. What is the difference in `Fare` between passengers that survived and passengers that did not survive?

In [9]:
# Question 1 
# Step 1: Slice a dataframe containing only male passengers that survived
male_survived = df[(df['Sex'] == 'male') & (df['Survived'] == 1)]

# Step 2: Use .median() and .mean() on appropriate columns to get values we're looking for
median_Parch = male_survived['Parch'].median()
mean_fare = male_survived['Fare'].mean()

print('Median Parch value of men that survived: {}'.format(median_Parch))
print('Mean fare value of men that survived: {}'.format(mean_fare))

Median Parch value of men that survived: 0.0
Mean fare value of men that survived: 40.82148440366974


**_Expected output for Question 1_**:

Median Parch value of men that survived: 0.0  
Mean fare value of men that survived: 40.82148440366974

In [10]:
# Question 2
# Step 1: Slice a dataframe containing only female passengers that did not survive. 
females_died = df[(df['Survived'] == 0) & (df['Sex'] == 'female')]

# Step 2: Get the count of how many total female passengers did not survive
total_females_died = len(females_died)

# Step 3: Get the count of females that died and had a Pclass of 2 or 3
females_died_pclass_2_3 = len(females_died[females_died['Pclass'] != 1])

# Step 4: Calculate percentage of females that died that had Pclass 2 or 3.
percentage = females_died_pclass_2_3 / total_females_died

print("Percentage of females that died that were Pclass 2 or 3: {}%".format(percentage * 100))

Percentage of females that died that were Pclass 2 or 3: 96.29629629629629%


**_Expected output for Question 2_**:

Percentage of females that died that were Pclass 2 or 3: 96.29629629629629%

In [12]:
# Question 3
# Step 1: Separate passengers that did and didn't survive into separate dataframes
survived_df = df[df['Survived'] == 1]
died_df = df[df['Survived'] == 0]

# Step 2: Get mean fare for each
mean_fare_survived = survived_df['Fare'].mean()
mean_fare_died = died_df['Fare'].mean()

# Step 3: Calculate the difference between mean_fare_survived and mean_fare_died
difference_in_fare = mean_fare_survived - mean_fare_died

print("Mean fare of passengers that survived: ${:.2f}".format(mean_fare_survived))
print("Mean fare of passengers that did not survive: ${:.2f}".format(mean_fare_died))
print("Difference in fare betweens survivors and non-survivors: ${:.2f}".format(difference_in_fare))

Mean fare of passengers that survived: $48.40
Mean fare of passengers that did not survive: $22.12
Difference in fare betweens survivors and non-survivors: $26.28


**_Expected output for Question 3_**:

Mean fare of passengers that survived: \$48.40   
Mean fare of passengers that did not survive: \$22.12   
Difference in fare betweens survivors and non-survivors: \$26.28 

### Summary

In this notebook, we learned about how to read in, store, and manipulate data in **_pandas_** using **_dataframes_**. Specifically, we learned how to:
* Read data from csv files into a dataframe
* How to access individual columns through dataframe slicing
* How to using dataframe slicing with conditional logic to get subgroups of data that are interesting to us
* Use descriptive statistics on data in pandas to help us analyze our data

We also briefly reviewed some major concepts in descriptive statistics:
* Measures of Center--mean, median, and mode
* Measures of Spread--standard deviation and variance
* Minimum, maximum, and quartiles

